### ISSUE TITLE
ability to include arbitrary C source text in @cImport blocks

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#2070]
Inspired by https://github.com/ziglang/zig/issues/1085#issuecomment-473390270

Here's a use case:

```zig
const c = @cImport({
    @cInclude("curses.h");
    @cSource(
        \\static const int MY_A_STANDOUT = A_STANDOUT;
    );
});

// now we have access to A_STANDOUT even though it was
// unable to be automatically translated
```

This would be incredibly easy to implement. It's just a question of whether it's a good idea.

And then there's the follow-up question of, do we really need `@cInclude`, `@cDefine`, and `@cUndef` if we have the all-powerful `@cSource`?

**Comment by Akuli** [comment]
+1. This might be a particularly good solution to #1085 because it could be used to solve a bunch of other problems that nobody can think of yet. For example, if there's something that you can do in C but not in Zig, you could define a C function to do the thing and then call that from Zig (I think).

**Comment by andrewrk** [comment]
Note that you can already do something very similar with status quo:

```zig
const c = @cImport({
    @cInclude("curses.h");
    @cInclude("mycode.h");
});
```

mycode.h:
```c
static const int MY_A_STANDOUT = A_STANDOUT;
```

**Comment by Akuli** [comment]
To `@cInclude` with a `build.zig` project, this can be added to `build.zig`:

```zig
    exe.addIncludeDir(".");
```

Now `@cInclude("src/mycode.h")` should work.

**Comment by daurnimator** [comment]
I think this feature would be really useful.

> And then there's the follow-up question of, do we really need `@cInclude`, `@cDefine`, and `@cUndef` if we have the all-powerful `@cSource`?

We wouldn't need them as builtins, but could have them as functions in the std library instead.

**Comment by tauoverpi** [comment]
There are a hand full of languages that have taken this approach such as [mercury], [ats2] and would remove the need of creating separate C source files though I'd be concerned about debugger and editor support for the code contained. I guess a reason for would be the ability to replace C code function by function inline if you could embed all of it but that is also rather messy closer to the 40-60% replaced mark and would promote more C use in zig code as I see it.

[mercury]: https://mercurylang.org/information/doc-release/mercury_ref/Using-pragma-foreign_005fdecl-for-C.html#Using-pragma-foreign_005fdecl-for-C
[ats2]: http://ats-lang.sourceforge.net/DOCUMENT/INT2PROGINATS/HTML/x2095.html

**Comment by andrewrk** [comment]
Solved instead by #20630
